wagtail/admin/staticfiles.py

```diff
@@ -35,20 +36,31 @@
     VERSION_HASH = None
 
 
-def versioned_static(path):
-    """
-    Wrapper for Django's static file finder to append a cache-busting query parameter
-    that updates on each Wagtail version
-    """
-    # An absolute path is returned unchanged (either a full URL, or processed already)
-    if path.startswith(("http://", "https://", "/")):
-        return path
+if os.environ.get("WAGTAIL_FAIL_ON_VERSIONED_STATIC", "0") == "1":
 
-    base_url = static(path)
+    def versioned_static(path):
+        raise Exception(
```

Comment:
I'm not convinced this is the best way to test this... it seems too specific and don't really simulate the issue, which is doing `collectstatic` with `DEBUG=False`. If we don't have a similar setup in our CI, we potentially could run into similar issues that's not caused by `versioned_static` (e.g. the reason why we added `collectstatic` in our CI like #9002).

I've found a different approach that works by setting `DEBUG = False` via an environment variable, will make a separate PR now.

Reply:
Also, just to add: I agree it is nice to show a more helpful message like this when developers put `versioned_static` at startup, but the problem is they need to explicitly set `WAGTAIL_FAIL_ON_VERSIONED_STATIC` to `1`, and neither this setting nor `versioned_static` are documented. If it's documented, I wouldn't mind us having these :D

Reply:
This isn't really intended to be a test in the sense of replicating something that would happen on a real-world site - it's more of a quality check on the Wagtail codebase, like [test_migrations](https://github.com/wagtail/wagtail/blob/main/wagtail/tests/test_migrations.py), to ensure that CI fails loudly if this has been overlooked. It doesn't need to be documented, as it's not intended for developers to run it outside of the CI environment.

Defaulting to 1 wouldn't work, because that would break legitimate uses of `versioned_static` during runtime too. I don't think there's any good way to distinguish between valid and invalid uses of `versioned_static` in code, so the idea is that the CI runner enables the `WAGTAIL_FAIL_ON_VERSIONED_STATIC` flag for a command such as `manage.py check` or `manage.py collectstatic` that isn't expected to render any views - if this fails, we know it's an invalid use of `versioned_static`.